### ISSUE TITLE
crashing with 'attempt to subtract with overflow' in input_buffer

### ISSUE THREAD

**Issue by RJ** [cBournhonesque/lightyear#560]
had a couple of crashes like this:

```
thread 'Compute Task Pool (0)' panicked at lightyear-2cfb5e6660946fe3/356f61d/lightyear/src/inputs/leafwing/input_buffer.rs:188:53:
attempt to subtract with overflow
```

https://github.com/cBournhonesque/lightyear/blob/main/lightyear/src/inputs/leafwing/input_buffer.rs#L188

something like this maybe?

```rust
if tick < start_tick || tick > Tick(start_tick.wrapping_add_signed(self.buffer.len() as i16 - 1)) {
    return None;
}
```

**Comment by cBournhonesque** [comment]
thanks, yes i'll fix that

**Comment by cBournhonesque** [comment]
I don't think the `tick` addition is the problem,  it's already handled with wrapping_add: https://github.com/cBournhonesque/lightyear/blob/87806d0c9cb6fef22978cf7b170089e37711d329/lightyear/src/utils/wrapping_id.rs#L135-L135 

For some reason it seems to be the subtract, but I don't understand how it could be happening

**Comment by RJ** [comment]
hmm. this is in the 'sandbox' version of my client/server game, which although it adds the lightyear plugins and components, just like the client and server code, it doesnt ever start or connect. it's a local only testbed.

perhaps something that would normally be pruned or some maintenance tasks aren't running as a result

**Comment by RJ** [comment]
https://github.com/cBournhonesque/lightyear/blob/main/lightyear/src/inputs/leafwing/input_buffer.rs#L102 too, in `set()`

**Comment by cBournhonesque** [comment]
You still get this? Could you try adding logs so we see the values that trigger the crash?

**Comment by RJ** [comment]
i believe it's happening when buffer_len (u16) reaches `i16::MAX + 1`, then the cast `as i16` converts it to `i16::MIN`, which you can't subtract 1 from. (after approx. 10mins at 64 ticks/sec)

buffer normally would be pruned i'm guessing, and isn't because i've not started the client or server.

**Comment by cBournhonesque** [comment]
Ah I see, that's helpful thanks

**Comment by RJ** [comment]
I'm avoiding the crash by adding the following system when running under my local sandbox environment, which doesn't start the lightyear server or client, it just adds the plugins.

Not much of a bug in the end, since i'm doing something weird.

However, `pop` was `pub(crate)`, which i had to change to `pub` to make this possible. (want a small pr for that?)


```rust
app.add_systems(FixedPostUpdate, prune_input_buffer);


/// when running in sandbox mode, normal lightyear systems aren't running, so we have to
/// prune the input buffer manually.
fn prune_input_buffer(
    mut q: Query<
        &mut InputBuffer<PlayerActions>,
        With<Player>,
    >,
    tick_manager: Res<TickManager>,
) {
    for mut buffer in q.iter_mut() {
        buffer.pop(tick_manager.tick());
    }
}
```

**Comment by cBournhonesque** [comment]
> (want a small pr for that?)

Sure! I think it's ok to make things like that `pub`; I guess most users wouldn't use it but advances users might

**Comment by mvlabat** [comment]
I've just triggered the same panic on 0.17.

Could we consider adding the logic of the system that @RJ posted to the crate, which would run whenever neither client or server isn't started? In my case, I also add the plugins but I call `ClientCommands::connect_client` only when a user decides to join a multiplayer game.

I don't think we should treat this as an advanced or a niche case, I'd guess that any game that wants to support both singleplayer and multiplayer modes or has a server borwser would be doing the same.

**Comment by cBournhonesque** [comment]
Sure, feel free to open a PR! (I'm busy with job search currently so I don't have too much time)

I guess it would just be a system that runs when the client is not connected or the inputs are disabled, and it would disable clear the InputBuffer?
In general, I would like to not include anything in the buffer if the ActionState or Inputs are disabled. I remember tackling this but I was blocked by something, can't remember why ... But yea a good first step is not buffer inputs in the InputBuffer if the client is not connected

**Comment by gak** [comment]
I'm getting similar (0.23.0 w/ BEI), but I think I've put the game into an odd state where the client is sending inputs when the server has a different input context active. This is just one client connected to a dedicated server.

<details>
<summary>Backtrace</summary>

```
SERVER: The application panicked (crashed).
SERVER: Message:  attempt to subtract with overflow
SERVER: Location: /Users/gak/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/lightyear_inputs-0.23.0/src/input_buffer.rs:153

SERVER: 12: core::panicking::panic_const::panic_const_sub_overflow::h65e0bf1a991800f0
SERVER:     at <unknown source file>
SERVER: 13: lightyear_inputs::input_buffer::InputBuffer<T>::set_raw::h931a2d4bde96c3e7
SERVER:     at <unknown source file>
SERVER: 14: lightyear_inputs::input_message::ActionStateSequence::update_buffer::h4034ba993d874ebe
SERVER:     at <unknown source file>
SERVER: 15: lightyear_inputs::server::receive_input_message::{{closure}}::{{closure}}::h5962c4a41a115a4e
SERVER:     at <unknown source file>
SERVER: 16: core::iter::traits::iterator::Iterator::try_for_each::call::{{closure}}::h8068108b1caf3fde
SERVER:     at <unknown source file>
SERVER: 17: core::iter::adapters::map::map_try_fold::{{closure}}::he143149dd6159f8c
SERVER:     at <unknown source file>
SERVER: 18: core::iter::traits::iterator::Iterator::try_fold::h5212d0790d0a8222
SERVER:     at <unknown source file>
SERVER: 19: lightyear_inputs::server::receive_input_message::{{closure}}::hba1524f1bf41d42c
SERVER:     at <unknown source file>

```

</details>

**Comment by cBournhonesque** [comment]
Yes there was a bug on TickDelta (https://github.com/cBournhonesque/lightyear/pull/1183/files#diff-ed77441f164cf1a834a23c6a225f044058882e8896587c633d8cac2b08975716R213) that was causing wrong tick values to be computed, which would in turn fill up the buffer.

After that bug was fixed I haven't seen the error anymore

**Comment by cBournhonesque** [comment]
Closing for now until someone can reproduce
